### ISSUE TITLE
Tracking the changes related to RM integration

### ISSUE THREAD

**Issue by artpol84** [openpmix/openpmix#722]
@rhc54
https://github.com/open-mpi/ompi/issues/5177 is not a first time some API changes in PMIx went out of our scope and thus were not included in Slurm plugin. It is hard to extract them from overall amount of changes. 
On in most cases I think it obvious to the developer making a change that this will affect other RM's.
Can we have a mailing list or other channel that will help developers maintaining RM integration to be in sync with the new release. It is especially important because RMs have their own release schedule.

**Comment by artpol84** [comment]
@karasevb FYI

**Comment by rhc54** [comment]
It is more of an OMPI issue, seems to me, as we have no control over if/when they absorb a release. Especially true in their case as the opal/pmix framework adds another layer of abstraction on top of PMIx that can "hide" any changes we make. It can get complicated as you have our release schedule, the OMPI schedule, the distro's schedule for updating PMIx as well as OMPI, and the RM schedule - all of which are uncoordinated.

In terms of tracking PMIx changes, we probably need to do a better job in the NEWS of capturing these things. I'm not sure what a mailing list would do for us here, or even if the RM community would subscribe to it. All we could say is "hey, PMIx made a change in release x.y - no idea if/when someone might absorb it". Not sure how that helps? Do you have some other thoughts on how to use it?

**Comment by artpol84** [comment]
But if we aware about the change that assumes that RM should now use a different PMIX key name or alike, it would help.

**Comment by rhc54** [comment]
I don't see how it can help as you have no idea if/when OMPI will pick it up, or MPICH, or...so what can Slurm do about it? You can't use a different name if the other codes are using the old one - or is it your plan to just pass every key and let the libraries pick?

The problem in this case stemmed from a change in HWLOC. The v1 xml string is incompatible with the v2 xml string, and so we had to distinguish between them. This creates an issue for HWLOC users as neither side (Slurm or OMPI, in this case) knows what the other one might be using. Hence, we let the provider (Slurm) "tag" the string with the version used to generate it. IIRC, there is a backward compatibility flag or function in HWLOC that allows you to deal with v1 strings (you are out-of-luck the other way around).

Alternatively, you can just let the other end attempt to parse the string and see if it succeeds. I don't believe HWLOC v1 will fail or return a bad topology if given a v2 string - I think it can detect that it's the wrong version and just return an error. Something you'd need to check as you wouldn't want OMPI to crash when trying to parse it.

My point being: it isn't clear if the RM needs to make a change, or OMPI et al need to do so. Passing every possible key might cause problems on its own, so I'm not sure that really works (like I said, it's something you'd need to check). Still isn't clear to me what we would tell people, or who would listen, as it is all very squishy.

Perhaps the ultimate answer is the need to be testing these things prior to release of a package like OMPI and/or Slurm any time a new PMIx release becomes available. Just remember that the Slurm plugin has to support others beyond just OMPI, so you can't exclusively tailor it in that direction.

**Comment by artpol84** [comment]
Ok, but then it seems like adding V1 key is error-prone. Instead I would use the old key name for HWLOC v1 and add a new key-name for HWLOC v2. Can you clarify why v1 was added?

**Comment by rhc54** [comment]
I truly believe this conversation should be closed here and moved to the OMPI site as it doesn't have anything to do with PMIx. We didn't remove or change any existing attribute. We added several attributes at the request of, and in consultation with, the HWLOC maintainers to convey new information. IIRC, the concern was that if someone passes you the old generic key, you really don't know how that XML string was generated. The v2 function for generating the XML string didn't change from their v1 series, and so the caller may not realize they are sending you something different from what they used to send. Specifying unique keys for the different incompatible representations was the desired solution, and we simply published them.

The OMPI integration was done by multiple people spanning several months. It was discussed in their issues and PR lists, devel telecons, and release manager meetings. The most recent changes in that area were committed 9 months ago to master and moved to v3.1.0 shortly thereafter. Since that time, there have been multiple release candidates published for testing - as you know, v3.1.0 took _forever_ to be released.

So what I think you are actually pointing to is a couple of things:

* there is a "hole" in the OMPI-Slurm integration testing. There needs to be more diligent testing of master, and especially of release candidates, and they probably need to expand their tests. In this case, OMPI falls back to each proc doing its own discovery, so there is no failure being reported - probably need a way of directly testing, for example, that the hwloc shmem feature worked.

* the Slurm community (i.e., SchedMD) may need to consider setting up a coordination mechanism between it and the various programming library groups. As I said, it goes beyond OMPI - SchedMD will have to figure out how they coordinate changes to the single pmix plugin across its various users.

Neither of those is a PMIx issue, I'm afraid.

**Comment by artpol84** [comment]
Before we will close it, lets confirm if I understood you correctly.
* The whole point of PMIx as a standard is to provide an abstraction layer between RM and application. 
* From the RM side there is a Server-side API and the set of predefined keys to describe job-level information.
* Server-side API looks like a standard to me so far. 
* Keys considered to be optional and out of standard. And you are saying that RM and Application side need to negotiate about what key's to use. This violates the idea of standard IMO. 
* What I as RM-side integrator would like to see is a well-defined set of keys with clear definition of what they mean (no duplications), so I can retrieve and provide them without thinking whether or not app support them.

For example, HWLOC topology. You provided some explanation about topo keys, but it doesn't fully answer my question. So let me rephrase:
* We had `PMIX_LOCAL_TOPO` key that obviously was corresponding to HWLOC v1 according to the time it was implemented.
* Later two new keys was added: `PMIX_HWLOC_XML_V1` and `PMIX_HWLOC_XML_V2` to explicitly distinguish v1 and v2.

For me it doesn't make sense as `PMIX_LOCAL_TOPO` and `PMIX_HWLOC_XML_V1` are the same and introducing V1 key creates the confusion. 
`PMIX_LOCAL_TOPO` and `PMIX_LOCAL_TOPO_V2` (or `PMIX_HWLOC_XML_V2`  if you like) would be sufficient to add explicit support for HWLOC v2 without breaking backward compatibility. Why do we need V1?
@rhc54, can you specifically answer this question?

* What would application-side developer think when looking on those 3 keys? As it stands now, I wouldn't consider `PMIX_LOCAL_TOPO`. And this is what OMPI actually did breaking backward compatibility and introducing undesired performance hit that is not easy to detect.
* How should RM developer decide which key to expose without going and talking to all of those communities? PMIx suppose to eliminate the need to communicate between RM and app in terms of which particular key to use.

**Comment by rhc54** [comment]
I think you are making an assumption that was clear in your mind, but not necessarily globally accepted. Specifically, you state:
```
We had PMIX_LOCAL_TOPO key that obviously was corresponding to HWLOC v1 according to the time it was implemented.
```
This correlation isn't true - in fact, there are multiple packages out there that provide topology information, not just HWLOC. It is true that OMPI exclusively supports HWLOC, which is where I suspect you derive that statement - but that isn't true of everyone.

As we worked thru the issue of HWLOC v1 vs v2, the point was made that someone passing ```PMIX_LOCAL_TOPO``` from a non-HWLOC source could cause a client using HWLOC to actually segfault. Looking at the various packages, it rapidly became clear that a generic topology key really won't work, and the HWLOC team asked us to create HWLOC-specific keys to protect them. This was done via the process I previously described.

What wasn't done, and in hindsight should be done, was to mark ```PMIX_LOCAL_TOPO``` as "deprecated", and then to require that anyone providing a topology representation to define a key unique to their software. We'll have to address that separately.

As to OMPI breaking backward compatibility: again, you need to take that up with them. It isn't a PMIx issue.

Just as PMIx has experienced growing pains, so will the rest of the RM and programming library communities. Never before has the RM provided information directly to the application or library - the library always discovered and published its own info, which the RM blindly circulated for it.

As a standard, we cannot dictate to either side what they must do - we can only provide standardized definitions and APIs. If an RM decides to use their own custom topology software or some alternative package and therefore doesn't provide any HWLOC key, then the programming library needs to detect that missing data and either adjust to fill the gap, or talk to the RM provider about ways to resolve the mismatch (e.g., by defining a new key that exposes their topology).

Thus, OMPI should (IMO) be checking to see if topology information that it recognizes is being provided, and if they feel that is an error, it should report it as such - right now, it simply discovers the topology for itself in that scenario. This was considered a proper response as either way the app gains access to the topology info - so I'm not sure I understand your claim that it compromised performance.

Note that RMs are _always_ going to define their own keys - e.g., JSM has done so to control PMIx_Spawn. Apps that choose to utilize environment-specific keys are welcome to do so, recognizing that it will likely impact portability. It would be nice if the RMs _chose_ to work with the PMIx community to avoid such situations - but we cannot force them to do so (indeed, in the case of JSM, even one of our own members chose to go it alone, so we don't have much of a moral position to stand on here). We can only hope that the user community loudly objects and pushes them to work with us. Same is true for programming libraries.

So I think the bottom line we can say is that the API (both server and client) is relatively easy to "standardize", and that has been done. Attributes are a much tougher problem, and support for attributes even harder. There is no perfect answer, just continually working to try and move forward.

**Comment by rhc54** [comment]
I believe this can be closed - we can explore other options as needed
